### PR TITLE
Exp overflow fix

### PR DESCRIPTION
**Pull request by aetn23**
Move monster exp calculation to `MonsterDeath`. This is done to prevent a bug where monster exp could overflow on hell if the monster is Diablo (per #5067). 

I know that right now `exp` member could be removed from monster, but I am working on a PR that will do that, among other things.

EDIT: removed `exp` field from `monster` since it simplifies things (no need to calculate the `exp` while loading)

### CODE REVIEW

File: Source/monster.cpp [hunk 1]
```diff
@@ -3891,8 +3886,16 @@ void M_StartHit(Monster &monster, const Player &player, int dam)
 
 void MonsterDeath(Monster &monster, Direction md, bool sendmsg)
 {
+	int monsterExp = sgGameInitInfo.nDifficulty == DIFF_NIGHTMARE ? 2 * (monster.exp + 1000) : sgGameInitInfo.nDifficulty == DIFF_HELL ? 4 * (monster.exp + 1000)
+	                                                                                                                                   : monster.exp;
+	if (monster.isUnique()) {
+		monsterExp *= 2;
```
Comment (DakkJaniels) [inline Source/monster.cpp:3892]: Should this bug be fixed? Probably wasn't intended to run the difficulty adjustment again for unique monsters 🤷🏽‍♂️

Comment (qndel) [reply Source/monster.cpp:3892]: + you should leave old values for vanilla compatibility or monsters will give 0 xp when you load devx save in vanilla xD

**Comment by aetn23** [comment]
I know that, I will handle it in a separate PR that I am working on that will do that and remove some other monster's members

File: Source/monster.cpp (hunk 1 above)
Comment (DakkJaniels) [reply Source/monster.cpp:3891]: @qndel is this a response to my other comment? If so, that's why I was saying the old values (from the array) should be saved to the save file, but we don't need to reload them in DevX.

File: Source/monster.cpp [hunk 2]
```diff
@@ -3884,18 +3883,23 @@ void M_StartHit(Monster &monster, const Player &player, int dam)
 	M_StartHit(monster, dam);
 }
 
-void MonsterDeath(Monster &monster, Direction md, bool sendmsg)
+int CalculateMonsterExp(uint16_t baseExp, bool isUnique)
 {
-	int monsterExp = sgGameInitInfo.nDifficulty == DIFF_NIGHTMARE ? 2 * (monster.exp + 1000) : sgGameInitInfo.nDifficulty == DIFF_HELL ? 4 * (monster.exp + 1000)
-	                                                                                                                                   : monster.exp;
-	if (monster.isUnique()) {
+	int monsterExp = sgGameInitInfo.nDifficulty == DIFF_NIGHTMARE ? 2 * (baseExp + 1000) : sgGameInitInfo.nDifficulty == DIFF_HELL ? 4 * (baseExp + 1000)
```
Comment (DakkJaniels) [inline Source/monster.cpp:3888]: I'm fine either way, but would switch/case be more readable here?

File: Source/monstdat.cpp [hunk 3]
```diff
@@ -25,7 +25,7 @@ constexpr uint16_t Uniq(_unique_items item)
...
+// _monster_id   name,                                       assetsSuffix,        soundSuffix,          trnFile,           availability,                 width, image, hasSpecial, hasSpecialSound,   frames[6],                  rate[6],                  minDunLvl, maxDunLvl, level, hitPointsMinimum, hitPointsMaximum, ai,           abilityFlags,                          intelligence, toHit, animFrameNum, minDamage, maxDamage, toHitSpecial, animFrameNumSpecial, minDamageSpecial, maxDamageSpecial, armorClass, monsterClass, resistance,                                                resistanceHell,                                                  selectionType,     treasure,      expBase
```
Comment (DakkJaniels) [inline Source/monstdat.cpp:28]: shouldn't this stay as exp, since that's what it is called in MonsterData?

**Comment by aetn23** [comment]
I removed the function to calculate exp, and simply calculate it in the monster `exp` member, since it is more readable.

**Comment by aetn23** [comment]
If the general consensus is to fix the unique monster overflow bug, let me know and I will quickly fix it. If there is no consensus, maybe it's better to merge this to master, and start new PR with unique monster exp fix?

**Comment by ephphatha** [comment]
> If the general consensus is to fix the unique monster overflow bug, let me know and I will quickly fix it. If there is no consensus, maybe it's better to merge this to master, and start new PR with unique monster exp fix?

If you want to do a quick interim fix capping the exp value when performing the calculation that can potentially overflow is going to be far easier to review and reason about, might as well do that in a separate PR and we can leave this one for a proper fix.

File: Source/monster.h [hunk 4]
```diff
@@ -307,6 +307,29 @@ struct Monster { // note: missing field _mAFNum
 		return pgettext("monster", data().name);
 	}
 
+	int exp(_difficulty difficulty) const
+	{
+		int monsterExp = data().exp;
```
Comment (ephphatha) [inline Source/monster.h:312]: This should be unsigned, negative experience values aren't valid.

File: Source/loadsave.cpp [hunk 5]
```diff
@@ -1388,7 +1389,7 @@ void SaveMonster(SaveHelper *file, Monster &monster)
 	file->WriteLE<int8_t>(monster.whoHit);
 	file->WriteLE<int8_t>(monster.level);
 	file->Skip(1); // Alignment
-	file->WriteLE<uint16_t>(monster.exp);
+	file->WriteLE<uint16_t>(monster.exp(sgGameInitInfo.nDifficulty));
```
Comment (ephphatha) [inline Source/loadsave.cpp:1392]: Should take this opportunity to fix the overflow when writing to the save file.

```suggestion
	file->WriteLE<uint16_t>(static_cast<uint16_t>(std::min<unsigned>(std::numeric_limits<uint16_t>::max(), monster.exp(sgGameInitInfo.nDifficulty))));
```

File: Source/loadsave.cpp [hunk 6]
```diff
@@ -638,7 +639,7 @@ void LoadMonster(LoadHelper *file, Monster &monster)
 	monster.whoHit = file->NextLE<int8_t>();
 	monster.level = file->NextLE<int8_t>();
 	file->Skip(1); // Alignment
-	monster.exp = file->NextLE<uint16_t>();
+	file->Skip(2); // Skip exp - the monster used to have this field
```
Comment (ephphatha) [inline Source/loadsave.cpp:642]:
```suggestion
	file->Skip(2); // Skip exp - now calculated from monstdat when the monster dies
```

**Comment by aetn23** [comment]
> If you want to do a quick interim fix capping the exp value when performing the calculation that can potentially overflow is going to be far easier to review and reason about, might as well do that in a separate PR and we can leave this one for a proper fix.

Okay, I think it's better to come up with a proper fix in this PR then. 

It seems like removing double unique monster calculation is a way to go, per reasons @DakkJaniels mentioned.

File: Source/monster.h [hunk 7]
```diff
@@ -307,9 +307,9 @@ struct Monster { // note: missing field _mAFNum
 		return pgettext("monster", data().name);
 	}
 
-	int exp(_difficulty difficulty) const
+	uint32_t exp(_difficulty difficulty) const
 	{
-		int monsterExp = data().exp;
+		unit32_t monsterExp = data().exp;
```
Comment (ephphatha) [inline Source/monster.h]: `unsigned` is fine, this doesn't need to be a fixed width field. It'll eventually get added to `Player::experience` which is `uint32_t` but that's after a division and cap has been applied, and the fixed width type there is because it's a persistent data structure. For function return values it's not critical to specify the size.

**Comment by aetn23** [comment]
I removed double unique monster exp calculation and commented exp function

File: Source/monster.cpp [hunk 8]
```diff
@@ -3659,8 +3653,7 @@ void M_StartHit(Monster &monster, const Player &player, int dam)
 
 void MonsterDeath(Monster &monster, Direction md, bool sendmsg)
 {
-	if (!monster.isPlayerMinion())
```
Comment (qndel) [inline Source/monster.cpp:3662]: why did you remove this?

Comment (aetn23) [reply Source/monster.cpp:3662]: It was removed by accident during merging, it is now reverted to how it should be

File: Source/monster.h [hunk 9]
```diff
@@ -317,6 +317,29 @@ struct Monster { // note: missing field _mAFNum
 		return pgettext("monster", data().name);
 	}
 
+	/**
+	 * @brief Calculates monster's experience points.
+	 * Fetches base exp value from @p MonstersData array.
```
Comment (qndel) [inline Source/monster.h:322]:
```cpp
	 * Fetches base exp value from @p MonstersData array.
```
what's `@p` ?

Comment (aetn23) [reply Source/monster.h:322]: `@p` makes the word after it appear in typewriter font

**Review by qndel — APPROVED** [review]
